File: src/lib/ipc/git.ts
import { invoke } from "@tauri-apps/api/core";
import type {
  BaseOption,
  DiffLine,
  HunkSummary,
  SideBySidePayload,
} from "$lib/types";

/**
 * Returns whether `path` lives in a Git repository.
 * On mobile builds the underlying Tauri command isn't compiled in, so we
 * swallow the resulting "command not found" error and return false so the
 * Diff tab degrades gracefully.
 */
export async function gitIsRepo(path: string): Promise<boolean> {
  try {
    return await invoke<boolean>("git_is_repo", { path });
  } catch {
    return false;
  }
}

export async function gitListBases(
  path: string,
  currentText?: string,
): Promise<BaseOption[]> {
  return await invoke<BaseOption[]>("git_list_bases", { path, currentText });
}

export async function gitHunks(
  path: string,
  currentText: string,
  base?: string,
): Promise<HunkSummary[]> {
  return await invoke<HunkSummary[]>("git_hunks", { path, currentText, base });
}

export async function gitFullDiff(
  path: string,
  currentText: string,
  base?: string,
): Promise<DiffLine[]> {
  return await invoke<DiffLine[]>("git_full_diff", { path, currentText, base });
}

export async function gitSideBySide(
  path: string,
  currentText: string,
  base?: string,
): Promise<SideBySidePayload> {
  return await invoke<SideBySidePayload>("git_side_by_side", {
    path,
    currentText,
    base,
  });
}

/**
 * Pure text-vs-text diffs. Used by the "Compare with disk" path which
 * doesn't go through Git.
 */
export async function diffTextHunks(
  oldText: string,
  newText: string,
): Promise<HunkSummary[]> {
  return await invoke<HunkSummary[]>("diff_text_hunks", { oldText, newText });
}

export async function diffTextFull(
  oldText: string,
  newText: string,
): Promise<DiffLine[]> {
  return await invoke<DiffLine[]>("diff_text_full", { oldText, newText });
}

export async function diffTextSideBySide(
  oldText: string,
  newText: string,
): Promise<SideBySidePayload> {
  return await invoke<SideBySidePayload>("diff_text_side_by_side", {
    oldText,
    newText,
  });
}

File: src-tauri/src/commands/git.rs
use std::path::PathBuf;

use mddiff_core::diff::{DiffLine, HunkSummary};
use mddiff_core::git::{BaseOption, SideBySidePayload, DEFAULT_BASE};

fn resolve(base: Option<String>) -> String {
    base.filter(|s| !s.trim().is_empty())
        .unwrap_or_else(|| DEFAULT_BASE.to_string())
}

#[tauri::command]
pub async fn git_is_repo(path: PathBuf) -> Result<bool, String> {
    Ok(mddiff_core::git::is_in_repo(&path))
}

#[tauri::command]
pub async fn git_list_bases(
    path: PathBuf,
    current_text: Option<String>,
) -> Result<Vec<BaseOption>, String> {
    mddiff_core::git::list_bases(&path, current_text.as_deref()).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn git_hunks(
    path: PathBuf,
    current_text: String,
    base: Option<String>,
) -> Result<Vec<HunkSummary>, String> {
    let base = resolve(base);
    mddiff_core::git::diff_text_against_base(&path, &current_text, &base).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn git_full_diff(
    path: PathBuf,
    current_text: String,
    base: Option<String>,
) -> Result<Vec<DiffLine>, String> {
    let base = resolve(base);
    mddiff_core::git::full_diff_against_base(&path, &current_text, &base).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn git_side_by_side(
    path: PathBuf,
    current_text: String,
    base: Option<String>,
) -> Result<SideBySidePayload, String> {
    let base = resolve(base);
    mddiff_core::git::side_by_side_against_base(&path, &current_text, &base)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn git_read_at(path: PathBuf, revspec: String) -> Result<String, String> {
    mddiff_core::git::read_at(&path, &revspec).map_err(|e| e.to_string())
}
